software/controller/lib/hal/i2c.cpp

```diff
@@ -0,0 +1,448 @@
+  EnableInterrupt(InterruptVector::I2C1_ERR, IntPriority::LOW);
+  EnableInterrupt(InterruptVector::DMA2_CH6, IntPriority::LOW);
+  EnableInterrupt(InterruptVector::DMA2_CH7, IntPriority::LOW);
+
+  // init i2c1
+  i2c1.Init(I2C1_BASE, DMA2_BASE, I2C::Speed::kFast);
+}
+
+// Those interrupt service routines are specific to our configuration, unlike
+// the I2C::Channel::*ISR() which are generic ISR associated with an I²C channel
+void I2C1_EV_ISR() { i2c1.I2CEventHandler(); };
+
+void I2C1_ER_ISR() { i2c1.I2CErrorHandler(); };
+
+void DMA2_CH6_ISR() { i2c1.DMAIntHandler(DMA_Chan::C6); };
+
+void DMA2_CH7_ISR() { i2c1.DMAIntHandler(DMA_Chan::C7); };
+#endif // BARE_STM32
+
+bool I2C::Channel::SendRequest(const Request &request) {
+  *(request.processed) = false;
+  // Queue the request if possible: check that there is room in the index buffer
+  if (buffer_.FreeCount() <= 0) {
+    return false;
+  }
+  // Add the current queue_ index into the buffer
+  if (buffer_.Put(ind_queue_)) {
+    queue_[ind_queue_] = request;
+  } else {
+    return false;
+  }
+
+  // In case of a write request, copy data to our write buffer
+  if (request.direction == ExchangeDirection::kWrite) {
+    if (!CopyDataToWriteBuffer(request.data, request.size)) {
+      return false;
+    }
+    // update the request's data pointer to the write buffer instead of the
+    // caller's scope variable
+    queue_[ind_queue_].data = &write_buffer_[write_buffer_index_];
+
+    // update the write buffer index
+    write_buffer_index_ += request.size;
+  }
+  // increment ind_queue_, which is the index at which the next request will be
+  // put in the queue, with wrapping around the queue.
+  if (++ind_queue_ >= kQueueLength) {
+    ind_queue_ = 0;
+  }
+
+  if (!transfer_in_progress_) {
+    StartTransfer();
+  }
+  // if the bus is busy, the transfer will be initiated by the interrupt
+  // handlers, our work is done!
+  return true;
+}
+
+bool I2C::Channel::CopyDataToWriteBuffer(const void *data,
+                                         const uint16_t size) {
+  // Check if the empty space at the end of the buffer is big enough to
+  // store all of the data
+  if (write_buffer_index_ + size > kWriteBufferSize) {
+    // It isn't ==> Check if the empty space at the beginning of the buffer
+    // is big enough to store all of the data instead
+    if (size >= write_buffer_start_) {
+      // There is no contiguous space left in buffer that is big enough, we
+      // can't safely send this Write request.
+      return false;
+    }
+
+    // We can write at the begining of the buffer, need to remember that we wrap
+    // at that index in order to properly wrap when updating write_buffer_start_
+    // when the (previous) transfer will end.
+    wrapping_index_ = write_buffer_index_;
+    write_buffer_index_ = 0;
+  }
+
+  memcpy(&write_buffer_[write_buffer_index_], data, size);
+  return true;
+}
+
+void I2C::Channel::StartTransfer() {
```

Comment:
I believe this is not an issue if we solve the possible race condition in SendRequest, since StartTransfer is a protected function that is only called from within the class, either in SendRequest and on specific interrupts that are (at least in theory) mutually exclusive. Will still try to solve this as it is probably doable and we want to be on the safe side...

Reply:
:+1: